go/kbfs/libfs/leveldb_storage.go

```diff
@@ -0,0 +1,681 @@
+// Copyright (c) 2012, Suryandaru Triandana <syndtr@gmail.com>
+// All rights reservefs.
+//
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This is a modified version of
+// github.com/syndtr/goleveldb/leveldb/storage/file_storage.go.
+//
+// Modifications: Copyright 2019 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libfs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	billy "gopkg.in/src-d/go-billy.v4"
+)
+
+var (
+	errReadOnly = errors.New("leveldb/storage: storage is read-only")
+)
+
+type levelDBStorageLock struct {
+	fs *levelDBStorage
+}
+
+func (lock *levelDBStorageLock) Unlock() {
+	if lock.fs != nil {
+		lock.fs.mu.Lock()
+		defer lock.fs.mu.Unlock()
+		if lock.fs.slock == lock {
+			lock.fs.slock = nil
+		}
+	}
+}
+
+type int64Slice []int64
+
+func (p int64Slice) Len() int           { return len(p) }
+func (p int64Slice) Less(i, j int) bool { return p[i] < p[j] }
+func (p int64Slice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+const logSizeThreshold = 1024 * 1024 // 1 MiB
+
+// levelDBStorage is a billy-filesystem-backed storage.
+type levelDBStorage struct {
+	fs       billy.Filesystem
+	readOnly bool
+
+	mu      sync.Mutex
+	flock   billy.File
+	slock   *levelDBStorageLock
+	logw    billy.File
+	logSize int64
+	buf     []byte
+	// Opened file counter; if open < 0 means closed.
+	open int
+	day  int
+}
+
+var _ storage.Storage = (*levelDBStorage)(nil)
+
+// OpenLevelDBStorage returns a new billy-filesystem-backed storage
+// implementation of the levelDB storage interface. This also acquires
+// a file lock, so any subsequent attempt to open the same path will
+// fail.
+func OpenLevelDBStorage(bfs billy.Filesystem, readOnly bool) (
+	s storage.Storage, err error) {
+	flock, err := bfs.OpenFile("LOCK", os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			flock.Close()
+		}
+	}()
+	err = flock.Lock()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		logw    billy.File
+		logSize int64
+	)
+	if !readOnly {
+		logw, err = bfs.OpenFile("LOG", os.O_WRONLY|os.O_CREATE, 0644)
+		if err != nil {
+			return nil, err
+		}
+		logSize, err = logw.Seek(0, os.SEEK_END)
```

Comment:
I know this is pre-existing, but curious how is this different from getting size from `bfs.Stat("LOG")` then get `.Size()`

Reply:
Not sure, it might just be due to a translation from C.  It might also be very slightly more efficient since it doesn't copy around all the other data in the stat struct.